cpp/src/sort/rank.cu

```diff
@@ -0,0 +1,220 @@
+        auto it = thrust::make_transform_iterator(
+            thrust::make_counting_iterator<size_type>(0), conv);
+        thrust::inclusive_scan(rmm::exec_policy(stream)->on(stream), it,
+                               it + size,
+                               dense_rank_sorted.data().get());
+      }
+    }
+
+    switch (method) {
+    case rank_method::FIRST:
+      // stable sort order ranking (no ties)
+      thrust::scatter(
+          rmm::exec_policy(stream)->on(stream),
+          thrust::make_counting_iterator<double>(1),
+          thrust::make_counting_iterator<double>(size + 1),
+          sorted_order_view.begin<size_type>(), rank_data);
+      break;
+    case rank_method::DENSE:
+      // All equal values have same rank and rank always increases by 1 between
+      // groups
+      thrust::scatter(rmm::exec_policy(stream)->on(stream),
+                      dense_rank_sorted.begin(), dense_rank_sorted.end(),
+                      sorted_order_view.begin<size_type>(), rank_data);
+      break;
+    case rank_method::MIN: {
+      // min  of first in the group
+      // All equal values have min of ranks among them.
+      rmm::device_vector<double> min_sorted(size, 0);
+      // algorithm: reduce_by_key(dense_rank, 1, n, min)
+      thrust::reduce_by_key(
+          rmm::exec_policy(stream)->on(stream), dense_rank_sorted.begin(),
+          dense_rank_sorted.end(), thrust::make_counting_iterator<double>(1),
+          thrust::make_discard_iterator(), min_sorted.begin(),
+          thrust::equal_to<double>{}, thrust::minimum<double>{});
+      auto sorted_min_rank = thrust::make_transform_iterator(
+          dense_rank_sorted.begin(), [min_rank = min_sorted.begin()] 
+          __device__( auto i) -> double { return min_rank[i - 1]; });
+      thrust::scatter(rmm::exec_policy(stream)->on(stream), sorted_min_rank,
+                      sorted_min_rank + size,
+                      sorted_order_view.begin<size_type>(), rank_data);
+    } break;
+    case rank_method::MAX: {
+      // max  of first in the group
+      // All equal values have max of ranks among them.
+      rmm::device_vector<double> max_sorted(size, 0);
+      // algorithm: reduce_by_key(dense_rank, 1, n, max)
+      thrust::reduce_by_key(
+          rmm::exec_policy(stream)->on(stream), dense_rank_sorted.begin(),
+          dense_rank_sorted.end(), thrust::make_counting_iterator<double>(1),
+          thrust::make_discard_iterator(), max_sorted.begin(),
+          thrust::equal_to<double>{}, thrust::maximum<double>{});
+      auto sorted_max_rank = thrust::make_transform_iterator(
+          dense_rank_sorted.begin(), [max_rank = max_sorted.begin()] 
+          __device__( auto i) -> double { return max_rank[i - 1]; });
+      thrust::scatter(rmm::exec_policy(stream)->on(stream), sorted_max_rank,
+                      sorted_max_rank + size,
+                      sorted_order_view.begin<size_type>(), rank_data);
```

Comment:
I find this code really hard to read, some spaces in between function calls and **consistent indenting** would be nice:
```cpp
      thrust::reduce_by_key(
          rmm::exec_policy(stream)->on(stream), dense_rank_sorted.begin(),
          dense_rank_sorted.end(), thrust::make_counting_iterator<double>(1),
          thrust::make_discard_iterator(), max_sorted.begin(),
          thrust::equal_to<double>{}, thrust::maximum<double>{});

      auto sorted_max_rank = thrust::make_transform_iterator(
          dense_rank_sorted.begin(), [max_rank = max_sorted.begin()] 
          __device__( auto i) -> double { return max_rank[i - 1]; });

      thrust::scatter(rmm::exec_policy(stream)->on(stream), sorted_max_rank,
          sorted_max_rank + size,
          sorted_order_view.begin<size_type>(), rank_data);
```
Personally, i really like this style below (but clang-format will destroy this anyway, so dont worry about it):
```cpp
      thrust::reduce_by_key(
          rmm::exec_policy(stream)->on(stream), 
          dense_rank_sorted.begin(),
          dense_rank_sorted.end(), 
          thrust::make_counting_iterator<double>(1),
          thrust::make_discard_iterator(), 
          max_sorted.begin(),
          thrust::equal_to<double>{}, 
          thrust::maximum<double>{});

      auto sorted_max_rank = thrust::make_transform_iterator(
          dense_rank_sorted.begin(), 
          [max_rank = max_sorted.begin()] 
          __device__( auto i) -> double { return max_rank[i - 1]; });

      thrust::scatter(
          rmm::exec_policy(stream)->on(stream), 
          sorted_max_rank,
          sorted_max_rank + size,
          sorted_order_view.begin<size_type>(), 
          rank_data);
```